graphql_compiler/query_pagination/__init__.py

```diff
@@ -0,0 +1,134 @@
+    filters with which the number of results can be constrained. Parameters for these filters are
+    then generated such that one of the new queries will return roughly a page of results, while the
+    other query will generate the rest of the results. This ensures that the two new queries' result
+    data is equivalent to the original query's result data.
+
+    Args:
+        schema_graph: SchemaGraph instance.
+        statistics: Statistics object.
+        query_ast: Document, AST of the GraphQL query that will be split.
+        parameters: dict, parameters with which query will be estimated.
+        num_pages: int, number of pages to split the query into.
+
+    Returns:
+        PaginationQueries namedtuple, describing two queries: the first is described in
+        next_page_query and next_page_parameters, which when executed will return roughly a page of
+        results of the original query; while the second query is described in continuation_query and
+        continuation_parameters, which will return the rest of the results of the original query.
+        The union of the two queries' result data is equivalent to the given query and parameter's
+        result data. There are no guarantees on the order of the result rows for the two generated
+        queries.
+    """
+    if num_pages <= 1:
+        raise AssertionError(u'Could not split query {} into pagination queries for the next page'
+                             u' of results, as the number of pages {} must be greater than 1: {}'
+                             .format(query_ast, num_pages, parameters))
+
+    parameterized_queries = generate_parameterized_queries(
+        schema_graph, statistics, query_ast, parameters
+    )
+
+    hydrated_parameters = hydrate_parameters_of_parameterized_query(
+        schema_graph, statistics, parameterized_queries, num_pages
+    )
+
+    result_pagination = PaginationQueries(
+        parameterized_queries.next_page_query,
+        hydrated_parameters.next_page_parameters,
+        parameterized_queries.continuation_query,
+        hydrated_parameters.continuation_parameterst,
+    )
+
+    return result_pagination
+
+
+def paginate_query(schema_graph, statistics, query_ast, parameters, page_size):
+    """Return two query ASTs usable for pagination whose union is equivalent to the given query.
+
+    The first query when executed will return approximately page_size results, while the second
+    query will return the rest of the results. In case the given query is estimated to return less
+    than a page of data, the query will not be split.
+    Since the cost estimator may underestimate or overestimate the actual number of pages, you
+    should expect the actual number of results to be within two orders of magnitude of the estimate.
+
+    Args:
+        schema_graph: SchemaGraph instance.
+        statistics: Statistics object.
+        query_ast: Document, AST of the GraphQL query that is being paginated.
```

Comment:
I think we'll always start with a graphql string, then parse it, then pass the AST in. Might be worth changing this to take in a string, which will also make the function more user-friendly, as a public facing method in `__init__.py`.

Reply:
Handing in the AST is preferred, as pagination may be called 100 times, and we'd have to pay the price of printing and parsing 100 times.
This is the argument @obi1kenobi put forth. Ultimately, this is an ease-of-use versus performance trade-off that you should probably speak to @obi1kenobi about.

Reply:
My suggestion would be to have a public function like `paginate_query_ast` that takes an AST and allows power users to reuse a pre-parsed AST, and separately have another public helper function called `paginate_query` that takes a query string, parses it, and then calls the other function. We've used this pattern successfully elsewhere in the compiler, and it seems to give us the best of both worlds at relatively minor cost.